### PR TITLE
move report file generation to library

### PR DESCRIPTION
**Pull request by fernflower**
- json report generation moved to leapp.utils.report;
- txt report is generated as part of leapp upgrade/
  leapp preupgrade command;
- txt report generation in VerifyCheckActor will be
  made obsolete and removed in leapp-repository PR as
  well as actor's library;
- remove leapp report command (might be reverted).

### CODE REVIEW

File: leapp/cli/upgrade/__init__.py [hunk 1]
```diff
@@ -211,11 +210,12 @@ def preupgrade(args):
...
     report_txt, report_json = [os.path.join(get_config().get('report', 'dir'),
                                             'leapp-report.{}'.format(f)) for f in ['txt', 'json']]
-    report_info([report_txt, report_json], fail=workflow.errors)
     # fetch all report messages as a list of dicts
-    messages = fetch_upgrade_report_raw(context, renderers=False)
-    with open(report_json, 'w+') as f:
-        json.dump({'entries': messages}, f, indent=2)
+    messages = fetch_upgrade_report_messages(context)
+    generate_report_file(messages, report_json)
+    # XXX FIXME original actor filtered all non-high messages leaving high/indicator only. Should we do the same?
```
Comment (Rezney) [inline leapp/cli/upgrade/__init__.py]: Definitely not. This has been a mistake which we need to fix...

File: leapp/utils/report.py [hunk 2]
```diff
@@ -1,7 +1,6 @@
 import json
-import six
 
-from jinja2 import Template
```
Comment (Rezney) [inline leapp/utils/report.py]: After this we should also remove jinja2 dependencies from spec but not in this PR as it any change has to be synced with leapp-repository.

File: leapp/utils/report.py [hunk 3]
```diff
@@ -42,31 +39,23 @@ def fetch_upgrade_report_raw(context_id, renderers=True):
...
-    for msg_data in messages_raw:
-        template = Template(msg_data.pop('renderers').get(renderer))
-        messages_to_print.append(template.render(msg_data))
-
-    return messages_to_print
+def generate_report_file(messages_to_report, path):
+    if path.endswith(".txt"):
+        with open(path, 'w') as f:
+            for message in messages_to_report:
+                f.write('Severity: {}\n'.format(message['severity']))
```
Comment (Rezney) [inline leapp/utils/report.py]: I am wondering what about flags, audiences, tags, external links. Some of those does not need to be in the plaintext report, e.g. tags, but links should definitely be there.

Question is, how to best render those. Currently it seems that:

`f.write('Remediation: {}\n'.format(message['detail']['remediation']))`

will be more JSON-like as ['remediation'] will return list of remediations and remediations may contain another list of commands.

We discussed this briefly with @shaded-enmity, we probably does not need to take care of formatting commands but should avoid some deep structures in our plaintext report. 

But it is matter of discussion of course.

File: leapp/cli/upgrade/__init__.py (hunk 1 above)
Comment (fernflower) [reply leapp/cli/upgrade/__init__.py]: Thanks, will remove the XXX FIXME then. Now 2 files (txt and json) finally match

File: leapp/utils/report.py (hunk 3 above)
Comment (fernflower) [reply leapp/utils/report.py]: Sure, I'll add anything we consider necessary, for now as the initial commit I intended to keep the basic functionality as it is.

**Comment by drehak** [comment]
Just wondering, should we merge this before #544 or after?

**Comment by fernflower** [comment]
@drehak #544 should definitely be merged first imho

File: leapp/utils/report.py [hunk 4]
```diff
@@ -18,3 +18,18 @@ def fetch_upgrade_report_messages(context_id):
...
+
+
+def generate_report_file(messages_to_report, path):
+    if path.endswith(".txt"):
+        with open(path, 'w') as f:
+            for message in messages_to_report:
+                f.write('Severity: {}\n'.format(message['severity']))
+                f.write('Title: {}\n'.format(message['title']))
+                f.write('Summary: {}\n'.format(message['summary']))
+                if message.get('detail', {}).get('remediation'):
```
Comment (shaded-enmity) [inline leapp/utils/report.py]:
```suggestion
                if message.get('detail', {}).get('remediations'):
```

https://github.com/oamg/leapp/blob/master/leapp/reporting/__init__.py#L237

File: leapp/utils/report.py [hunk 5]
```diff
@@ -18,3 +18,18 @@ def fetch_upgrade_report_messages(context_id):
...
+                    f.write('Remediation: {}\n'.format(message['detail']['remediation']))
```
Comment (shaded-enmity) [inline leapp/utils/report.py]: Ditto from above, and apart from that `remediations` is a list of possible remediations in `{type: ..., value: ...}` format so we probably want to do something like this here
```
['{type}\n  {value}'.format(**r) for r in message['detail']['remediations']]
```

File: leapp/utils/report.py [hunk 6]
```diff
@@ -18,3 +18,19 @@ def fetch_upgrade_report_messages(context_id):
...
+
+def generate_report_file(messages_to_report, path):
+    if path.endswith(".txt"):
+        with open(path, 'w') as f:
+            for message in messages_to_report:
+                f.write('Severity: {}\n'.format(message['severity']))
+                f.write('Title: {}\n'.format(message['title']))
+                f.write('Summary: {}\n'.format(message['summary']))
+                if message.get('detail', {}).get('remediations'):
+                    f.write('Remediation: {}\n'.format(','.join(["{}: {}".format(r['type'], r['context'])
```
Comment (shaded-enmity) [inline leapp/utils/report.py]: Ahh, so when `type == command` we get this:
```
[u'yum', u'remove', u'leapp leapp-deps leapp-pes-data leapp-repository leapp-repository-deps leapp-repository-sos-plugin python2-leapp']
```
So I guess that we want something like `' '.join(...)` -- I think that we want to move this to a separate function like `_write_remediations()` for readability

File: leapp/utils/report.py [hunk 7]
```diff
@@ -18,3 +18,19 @@ def fetch_upgrade_report_messages(context_id):
...
+                                                                 for r in message['detail']['remediations']])))
+                f.write('-' * 40 + '\n')
+    elif path.endswith(".json"):
+        with open(path, 'w') as f:
+            json.dump({'entries': messages_to_report}, f, indent=2)
```
Comment (shaded-enmity) [inline leapp/utils/report.py]: Can we also get here the UUID of the run as stored in DB?

```suggestion
            json.dump({'entries': messages_to_report, 'leapp_run_id': id}, f, indent=2)
```

File: leapp/utils/report.py (hunk 6 above)
Comment (fernflower) [reply leapp/utils/report.py]: I'd rather have a class method like repr() for each Remediation type than one function, but you are right

**Pull request merged**